libparsec/crates/client/src/workspace/transactions/file_operations.rs

```diff
@@ -0,0 +1,441 @@
+                    .unwrap_or_default(),
+            );
+            new_last_block_chunks.push(new_last_block_last_chunk);
+            new_last_block_chunks
+        };
+
+        // Remove all blocks after the new size, and also the last block...
+        let blocks_drain_at = last_block_index;
+        // ...but in the drain ignore the first chunks of the last block (as there are still in use)...
+        let chunks_skip_drain_until = last_block_last_chunk_index + 1;
+        // ...and finally we have an updated last block to append (containing the chunks ignored in the drain).
+        let maybe_new_last_block_chunks = Some(new_last_block_chunks);
+        (
+            blocks_drain_at,
+            chunks_skip_drain_until,
+            maybe_new_last_block_chunks,
+        )
+    };
+
+    // Update the manifest with the new last block
+
+    manifest.need_sync = true;
+    manifest.size = size;
+    manifest.updated = timestamp;
+    // Only keep the blocks that should still be used as-is...
+    let removed_chunks = manifest
+        .blocks
+        .drain(blocks_drain_at..)
+        .flatten()
+        .skip(chunks_skip_drain_until)
+        .map(|x| x.id)
+        .collect();
+    // ...then re-insert the last block if only part of it is still used
+    if let Some(new_last_block_chunks) = maybe_new_last_block_chunks {
+        manifest.blocks.push(new_last_block_chunks);
+    }
+
+    removed_chunks
+}
+
+/// Prepare a resize operation by updating the provided manifest.
+///
+/// Return a `Vec` of write operations that must be performed in order for the updated manifest to become valid.
+/// Each write operation consists of a new chunk to store, along with an offset to apply to the corresponding raw data.
+/// Note that the raw data also needs to be sliced to the chunk size and padded with null bytes if necessary.
+/// Also return a list of chunk IDs that must cleaned up from the storage, after the updated manifest has been successfully stored.
+pub(crate) fn prepare_resize(
+    manifest: &mut LocalFileManifest,
+    size: u64,
+    timestamp: DateTime,
+) -> (Vec<WriteOperation>, Vec<ChunkID>) {
+    if size >= manifest.size {
+        // Extend
+        prepare_write(manifest, 0, size, timestamp)
+    } else {
+        // Truncate
+        let removed_ids = prepare_truncate(manifest, size, timestamp);
+        (vec![], removed_ids)
+    }
+}
+
+// Prepare reshape
+
+pub(crate) struct ReshapeBlockOperation<'a> {
+    destination_chunks: &'a mut Vec<Chunk>,
+    pub source: Vec<Chunk>,
+}
+
+impl ReshapeBlockOperation<'_> {
+    pub fn destination(&self) -> &Chunk {
+        &self.destination_chunks[0]
+    }
+
+    pub fn rollback(self) {
+        *self.destination_chunks = self.source;
+    }
+
+    pub fn commit(self, chunk_data: &[u8]) {
+        self.destination_chunks[0]
+            .promote_as_block(chunk_data)
+            .expect("chunk is block-compatible");
+    }
```

Comment:
Maybe move `mem::replace` here to remove rollback method?

Reply:
:+1: fixed